File: apps/api/app/schemas/gamification.py
from datetime import date, datetime
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, Field


class RewardOut(BaseModel):
    xp_gained: int
    coins_gained: int
    total_xp: int
    level: int
    leveled_up: bool
    current_streak: int
    streak_increased: bool
    freeze_used: bool
    goal_reached: bool
    new_achievements: List[str]


class StatsOut(BaseModel):
    xp: int
    level: int
    xp_into_level: int
    xp_for_next_level: int
    coins: int
    total_reviews: int
    current_streak: int
    longest_streak: int
    streak_freezes: int
    daily_goal: int
    reviews_today: int
    goal_reached_today: bool
    league_tier: str
    league_tier_index: int


class AchievementOut(BaseModel):
    code: str
    category: str
    xp_reward: int
    coin_reward: int
    unlocked: bool
    unlocked_at: Optional[datetime] = None


class DailyGoalUpdate(BaseModel):
    daily_goal: int = Field(ge=5, le=200)


class FreezePurchaseOut(BaseModel):
    streak_freezes: int
    coins: int


class LeaderboardMember(BaseModel):
    rank: int
    user_id: UUID
    display_name: str
    avatar_url: Optional[str] = None
    xp: int
    is_me: bool


class LeaderboardOut(BaseModel):
    tier: str
    tier_index: int
    iso_week: str
    promote_top: int
    relegate_bottom: int
    my_rank: int
    members: List[LeaderboardMember]


class HeatmapDay(BaseModel):
    day: date
    reviews_count: int
    xp_earned: int
    goal_reached: bool


class HeatmapOut(BaseModel):
    days: List[HeatmapDay]

File: apps 2/web/lib/gamification.ts
import { apiFetch } from "@/lib/api";

export interface Stats {
  xp: number;
  level: number;
  xp_into_level: number;
  xp_for_next_level: number;
  coins: number;
  total_reviews: number;
  current_streak: number;
  longest_streak: number;
  streak_freezes: number;
  daily_goal: number;
  reviews_today: number;
  goal_reached_today: boolean;
  league_tier: string;
  league_tier_index: number;
}

export interface Reward {
  xp_gained: number;
  coins_gained: number;
  total_xp: number;
  level: number;
  leveled_up: boolean;
  current_streak: number;
  streak_increased: boolean;
  freeze_used: boolean;
  goal_reached: boolean;
  new_achievements: string[];
}

export interface Achievement {
  code: string;
  category: string;
  xp_reward: number;
  coin_reward: number;
  unlocked: boolean;
  unlocked_at: string | null;
}

export interface LeaderboardMember {
  rank: number;
  user_id: string;
  display_name: string;
  avatar_url: string | null;
  xp: number;
  is_me: boolean;
}

export interface Leaderboard {
  tier: string;
  tier_index: number;
  iso_week: string;
  promote_top: number;
  relegate_bottom: number;
  my_rank: number;
  members: LeaderboardMember[];
}

export interface DailyQuest {
  code: string;
  progress: number;
  target: number;
  xp_reward: number;
  completed: boolean;
  game_type: string;
  source_category: string | null;
}

export interface DailyQuests {
  day: string;
  game_xp_today: number;
  completed_count: number;
  total_count: number;
  quests: DailyQuest[];
}

/** Header/dashboard listen for this to refetch stats after a review session. */
export const STATS_CHANGED_EVENT = "words:stats-changed";
export const QUESTS_CHANGED_EVENT = "words:quests-changed";

export function notifyStatsChanged() {
  if (typeof window !== "undefined") {
    window.dispatchEvent(new Event(STATS_CHANGED_EVENT));
  }
}

export function notifyQuestsChanged() {
  if (typeof window !== "undefined") {
    window.dispatchEvent(new Event(QUESTS_CHANGED_EVENT));
  }
}

export const gamificationApi = {
  stats: () => apiFetch<Stats>("/me/stats", { auth: true }),

  dailyQuests: () => apiFetch<DailyQuests>("/me/daily-quests", { auth: true }),

  achievements: () => apiFetch<Achievement[]>("/me/achievements", { auth: true }),

  leaderboard: () => apiFetch<Leaderboard>("/leaderboard", { auth: true }),

  setDailyGoal: (dailyGoal: number) =>
    apiFetch<Stats>("/me/daily-goal", {
      method: "PUT",
      body: { daily_goal: dailyGoal },
      auth: true,
    }),

  buyStreakFreeze: () =>
    apiFetch<{ streak_freezes: number; coins: number }>("/me/streak-freeze", {
      method: "POST",
      auth: true,
    }),
};
